[14.0][FIX] account_financial_report: Prevent error related to currency from General ledger

victoralmau opened the pull request:
Related to: https://github.com/OCA/account-financial-reporting/issues/1039

Prevent error related to currency from General ledger

Use case: Generate report (showing foreign currency) with accounts with defined currency.

Traceback:

```
File "/odoo/odoo-server/odoo/addons/base/models/ir_qweb_fields.py", line 448, in value_to_html
fmt = "%.{0}f".format(display_currency.decimal_places)
AttributeError: 'int' object has no attribute 'decimal_places'

The above exception was the direct cause of the following exception:

Error to render compiling AST
AttributeError: 'int' object has no attribute 'decimal_places'
Template: account_financial_report.report_general_ledger_lines
Path: /t/div/div[2]/t[6]/t[1]/div[1]/t[1]/span/t
Node: <t t-raw="account_or_group_item_object['init_bal']['bal_curr']" t-options="{'widget': 'monetary', 'display_currency': account['currency_id']}"/>

```
Please @pedrobaeza can you review it?

@Tecnativa TT42804

victoralmau committed:
[FIX] account_financial_report: Prevent error related to currency from General ledger

Use case: Generate report (showing foreign currency) with accounts with defined currency.

Traceback:

File "/odoo/odoo-server/odoo/addons/base/models/ir_qweb_fields.py", line 448, in value_to_html
fmt = "%.{0}f".format(display_currency.decimal_places)
AttributeError: 'int' object has no attribute 'decimal_places'

The above exception was the direct cause of the following exception:

Error to render compiling AST
AttributeError: 'int' object has no attribute 'decimal_places'
Template: account_financial_report.report_general_ledger_lines
Path: /t/div/div[2]/t[6]/t[1]/div[1]/t[1]/span/t
Node: <t t-raw="account_or_group_item_object['init_bal']['bal_curr']" t-options="{'widget': 'monetary', 'display_currency': account['currency_id']}"/>

TT42804

GSLabIt reviewed (approved):
LGTM

pedrobaeza commented on account_financial_report/report/general_ledger.py:
```diff
@@ -648,18 +648,22 @@ def _create_general_ledger(
                     "grouped_by": grouped_by,
                 }
             )
+
```
Don't add empty lines

pedrobaeza reviewed (commented):
I don't quite get why you need to have currency record instead of the ID, and then why you call the dictionary key `currency_id` instead of `currency`.

I have the feeling that the browse should be done where such key is consumed, not on the data preparation. If not possible, then rename the key to just `currency`.

And please, as always requested, include the traceback that motivates this fix in the PR and in the commit message.

victoralmau commented:
> I don't quite get why you need to have currency record instead of the ID, and then why you call the dictionary key `currency_id` instead of `currency`.
> 
> I have the feeling that the browse should be done where such key is consumed, not on the data preparation. If not possible, then rename the key to just `currency`.

I have thought about defining a new key called currency but the diff needed in the report would be very large and could have errors if not changed everywhere.

> 
> And please, as always requested, include the traceback that motivates this fix in the PR and in the commit message.

Ok.

pedrobaeza commented:
I still insist that we should put order here. If not, sooner or later we will commit the same mistake, as seeing `currency_id` will provoke the reaction to put an ID, not a record. Please check all the places where the dictionary is built, and decide if you let `currency_id` and do the browse where that value is used, or you rename to `currency` the dictionary key.

victoralmau commented:
Changes done

ferran-S73 reviewed (approved):
Functional and code review

pedrobaeza commented on account_financial_report/report/abstract_report_xlsx.py:
```diff
@@ -531,7 +531,10 @@ def _get_currency_amt_format(self, line_object, report_data):
+            if isinstance(line_object["currency_id"], int):
+                currency = self.env["res.currency"].browse(line_object["currency_id"])
+            else:
+                currency = line_object["currency_id"]
```
I still don't get: if you have renamed the variable to `currency`, why are you loading here `line_object["currency_id"]`?

I also still see the duality `currency_id` / `currency`, and this is not the ideal, as we can't assure if both are available.

victoralmau replied:
I don't agree with making all these changes either.

Another possible approach (at template level only) could be:

Modify all records that use `currency_id` (https://github.com/OCA/account-financial-reporting/blob/14.0/account_financial_report/report/templates/general_ledger.xml#L324 for example) to something "_similar_" to https://github.com/OCA/account-financial-reporting/blob/14.0/account_financial_report/report/templates/general_ledger.xml#L586-L605 (define `account_currency` object and use it to display the value).

pedrobaeza replied:
You don't have to keep compatibility with both `currency_id` and `currency`. Just make it homogeneous across entire module, which was not previously.

victoralmau committed:
[FIX] account_financial_report: Prevent error related to currency from General ledger

Use case: Generate xlsx report (showing foreign currency) with accounts with defined currency.

Traceback:

File "/opt/odoo/auto/addons/account_financial_report/report/abstract_report_xlsx.py", line 535, in _get_currency_amt_format
    field_name = "{}_{}".format(field_prefix, currency.name)
AttributeError: 'int' object has no attribute 'name'

victoralmau commented:
New applied approach.

Now everything is separated into 2 commits with as few changes as possible (no need to create `currency` everywhere).

pedrobaeza commented:
/ocabot merge minor

whotopia commented:
Did doing this, break this?  https://github.com/OCA/account-financial-reporting/issues/1053
